src/main/java/org/spongepowered/api/command/args/GenericArguments.java

```diff
@@ -1585,4 +1628,259 @@ public Text getUsage(CommandSource src) {
         }
     }
 
+    /**
+     * Expect an argument to represent a {@link URL}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement url(Text key) {
+        return new URLElement(key);
+    }
+
+    private static class URLElement extends KeyElement {
+
+        protected URLElement(Text key) {
+            super(key);
+        }
+
+        @Nullable
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            String str = args.next();
+            URL url;
+            try {
+                url = new URL(str);
+            } catch (MalformedURLException ex) {
+                throw new ArgumentParseException(Text.of("Invalid URL!"), ex, str, 0);
+            }
+            try {
+                url.toURI();
+            } catch (URISyntaxException ex) {
+                throw new ArgumentParseException(Text.of("Invalid URL!"), ex, str, 0);
+            }
+            return url;
+        }
+    }
+
+    /**
+     * Expect an argument to return an IP address, in the form of an
+     * {@link InetAddress}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement ip(Text key) {
+        return new IpElement(key, false);
+    }
+
+    /**
+     * Expect an argument to return an IP address, in the form of an
+     * {@link InetAddress}, or if nothing matches and the source is a
+     * {@link RemoteSource}, return the source's address.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement ipOrSource(Text key) {
+        return new IpElement(key, true);
+    }
+
+    private static class IpElement extends KeyElement {
+
+        private final boolean self;
+        private final PlayerCommandElement possiblePlayer;
+
+        protected IpElement(Text key, boolean self) {
+            super(key);
+            this.self = self;
+            this.possiblePlayer = new PlayerCommandElement(key, false);
+        }
+
+        @Nullable
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            if (!args.hasNext() && this.self) {
+                if (source instanceof Player) {
```

Comment:
According to the javadoc for `ipOrSource` you meant `RemoteSource` here and in the cast below.